### PR TITLE
Testsuite: add configurable executables to DEAL_II_ADD_TEST

### PR DESCRIPTION
**Pull request by tamiko**
This pull request adds a second type of tests to the testsuite that consists of
parameter file "test.prm" with an output file "test.<...>.output". It
assumes that a cmake variable

  TEST_TARGET (or TEST_TARGET_DEBUG and TEST_TARGET_RELEASE)

is defined and will invoke said target with the full path to the parameter
file. (The first type of tests consist of a source file "test.cc" instead,
that is compiled and linked.)

TODO: Detailed documentation for this feature

### CODE REVIEW

File: cmake/macros/macro_deal_ii_add_test.cmake [hunk 1]
```diff
@@ -16,10 +16,28 @@
...
-# output of the compiled source file is compared against the file
-# comparison file.
+#
+# The testsuite distinguishes two different kinds of tests:
+#
+# - A combination of a source file "${test_name}.cc" (containing a main
+#   function) with a file "${comparison_file}" defines an executable that
+#   is compiled and linked against deal.II. Its output is compared with the
+#   comparison file. Additional libraries (like a library from a user
+#   project with code to test) the target should be link against can be
```
Comment (bangerth) [inline cmake/macros/macro_deal_ii_add_test.cmake]: link -> linked

File: cmake/macros/macro_deal_ii_add_test.cmake [hunk 2]
```diff
@@ -128,39 +139,65 @@ MACRO(DEAL_II_ADD_TEST _category _test_name _comparison_file)
...
+      ELSEIF(EXISTS "${CMAKE_CURRENT_SOURCE_DIR}/${_test_name}.prm")
+
+        IF(NOT "${TEST_TARGET_${_build}}" STREQUAL "")
+          SET(_target ${TEST_TARGET_${_build}})
+        ELSEIF(NOT "${TEST_TARGET}" STREQUAL "")
+          SET(_target ${TEST_TARGET})
+        ELSE()
+          MESSAGE(FATAL_ERROR
+            "\nFor ${_comparison_file}: \"${_test_name}.prm\" provided, but "
+            "no variable \"\${TEST_TARGET}\" defined.\n\n"
```
Comment (bangerth) [inline cmake/macros/macro_deal_ii_add_test.cmake]: It's not necessary to define TEST_TARGET. It is sufficient to define TEST_TARGET_DEBUG/RELEASE. Might want to adjust the error message.

File: cmake/scripts/run_test.sh [hunk 3]
```diff
@@ -40,7 +40,7 @@ run(){
   rm -f failing_output
   rm -f output
 
-  ${RUN_COMMAND} > stdout
+  ${RUN_COMMAND} > stdout 2>&1
```
Comment (bangerth) [inline cmake/scripts/run_test.sh]: This almost deserves its own patch as it is unrelated to the other stuff here.

**Comment by bangerth** [comment]
Looks mostly good. Want to add the documentation here, or in a separate patch?

File: cmake/scripts/run_test.sh (hunk 3 above)
Comment (tamiko) [inline cmake/scripts/run_test.sh]: Good point. Done.

**Comment by tamiko** [comment]
@bangerth I'd like to see this merged. The documentation I plan to write will be a bigger change.

File: cmake/macros/macro_deal_ii_add_test.cmake [hunk 4]
```diff
@@ -128,39 +138,66 @@ MACRO(DEAL_II_ADD_TEST _category _test_name _comparison_file)
...
+      ELSEIF(EXISTS "${CMAKE_CURRENT_SOURCE_DIR}/${_test_name}.prm")
+
+        IF(NOT "${TEST_TARGET_${_build}}" STREQUAL "")
+          SET(_target ${TEST_TARGET_${_build}})
+        ELSEIF(NOT "${TEST_TARGET}" STREQUAL "")
+          SET(_target ${TEST_TARGET})
+        ELSE()
+          MESSAGE(FATAL_ERROR
+            "\nFor ${_comparison_file}: \"${_test_name}.prm\" provided, but "
+            "neither \"\${TEST_TARGET}\", nor \"\${TEST_TARGET}_{DEBUG|RELEASE}"
```
Comment (tjhei) [inline cmake/macros/macro_deal_ii_add_test.cmake]: I think ```\${TEST_TARGET_${_build}}``` would be nice in the error message here.

File: cmake/scripts/run_test.sh (hunk 3 above)
Comment (tjhei) [inline cmake/scripts/run_test.sh]: What do you mean by "done"? So with this change stderr will no longer be reported if the test also generates "output"? That looks weird.

**Comment by tamiko** [comment]
@tjhei This is not part of this pull request any more.
